File: src/olcw78 (die yoong)/jsAlgorithmStudy/week02/08_괄호짝맞추기.js
/**
 * @param str {string}
 * @return {boolean}
 */
function solution(str) {
    let cnt = 0;
    for (const c of str) {
        // ( 로 시작 -> +1
        if (c === '(') {
            cnt += 1
            continue
        }

        // ( 아니면 -> ) -> -1
        cnt -= 1

        // 계속 진행하며 괄호가 닫힌게 더 많으면 실패
        if (cnt < 0) return false
    }

    // 마지막에 괄호쌍이 맞으면 성공
    return cnt === 0
}

const r1 = solution("(())()")
console.log(r1)

const r2 = solution("((())()")
console.log(r2)

const r3 = solution("(((()")
console.log(r3)

const r4 = solution(")()()(")
console.log(r4)

const r5 = solution("(())((()()()))")
console.log(r5)

File: src/olcw78 (die yoong)/jsAlgorithmStudy/week02/08_괄호짝맞추기.go
package week02

import "fmt"

func solution08(str string) bool {
	cnt := 0
	for _, c := range str {
		if c == '(' {
			cnt++
			continue
		}

		cnt--
		if cnt < 0 {
			return false
		}
	}

	return cnt == 0
}

func Run08() {
	r1 := solution08("(())()")
	fmt.Println(r1)

	r2 := solution08("()()")
	fmt.Println(r2)

	r3 := solution08("()()()")
	fmt.Println(r3)

	r4 := solution08("(((()")
	fmt.Println(r4)
}
